torch/distributed/device_mesh.py

```diff
@@ -317,41 +271,41 @@ def _get_slice_mesh_dims(
                     f"Valid mesh_dim_names are {valid_mesh_dim_names}."
                 )
 
-            # Validate the order of the slice mesh dim indices.
-            # This needs to be in ascending order.
-            curr_idx = -1
-            slice_mesh_dims = []
-            for mesh_dim_name in mesh_dim_names:
-                if mesh_dim_name in flatten_name_to_root_dims:
-                    mesh_indices = flatten_name_to_root_dims[mesh_dim_name]
-                    # TODO: this doesn't allow non-contiguous slicing with flatten dim yet. next_idx
-                    # should be mesh_indices[0] once we support non-contiguous slicing with flatten dim.
-                    next_idx = mesh_indices[-1]
-                    slice_mesh_dims.append(mesh_indices)
-                else:
-                    next_idx = device_mesh.mesh_dim_names.index(mesh_dim_name)
-                    slice_mesh_dims.append((next_idx,))
-                if next_idx <= curr_idx:
-                    raise KeyError(
-                        f"Invalid mesh_dim_names {mesh_dim_names} specified. "
-                        f"Found mesh dim indices to slice: {slice_mesh_dims}. "
-                        "Mesh dim indices should be in ascending order."
+            layout_sliced = []
+            for name in mesh_dim_names:
+                if name in device_mesh.mesh_dim_names:
+                    layout_sliced.append(
+                        device_mesh._layout[device_mesh.mesh_dim_names.index(name)]
                     )
-                curr_idx = next_idx
+                elif name in flatten_name_to_root_layout:
```

Comment:
Maybe this will get addressed in later PRs, but what I see here is that we make a distinction between two "types" of dimensions: the ones that come from flattening and all the other ones. This distinction feels artificial to me. Once a dimension has been flattened, we compute its (nested?) sizes and strides, just like any other dimension, and then it should become indistinguishable and we should have a single code path. Why does it still require special handling?

Reply:
I agree with @ezyang, it'd be very valuable to get a clear description of the legacy constraints (documented in the code, not as GitHub comments) so we can keep them in mind when proposing changes.

As I understand it, the legacy behavior you're describing occurs during lookup of a named mesh dim (i.e., inside the `__getitem__` method): on the _root_ mesh (and only on that one!) needs to be aware of dimensions that are not part of the original dims at creation, but which were obtained via flattening some of them. Is that so?

If that's the case then I'd expect to see some ad-hoc code inside the `__getitem__` method (plus some bookkeeping elsewhere), but the rest of the code to be relatively unimpacted.

Reply:
Yes it is because we need `__getitem__` to work with flatten for BC reason. And to clarify, this requires the bookeeping to have hack both at flatten and `__getitem__`.